Merge 2.6.x -> 3.0.x

Magisus opened the pull request.

chrisleicester committed:
(CDPE-1813) Add module deploy info to .r10k-deploy.json file
With this change, module information will be added to the output file '.r10k-deploy.json' for external consumption (including the module name, version, and git sha associated with the commit used if pulling the module from a git repository

Magisus committed:
(CDPE-1813) Add module deploy info to .r10k-deploy.json file (#955)

(CDPE-1813) Add module deploy info to .r10k-deploy.json file

reidmv committed:
Add test for forge module to .r10k-deploys suite

This additional test ensures that a Forge module (which does not have a
Git repo) is written as expected to the .r10k-deploys file.

Magisus committed:
Add test for forge module to .r10k-deploys suite (#961)

Add test for forge module to .r10k-deploys suite

chrisleicester commented on lib/r10k/action/deploy/environment.rb:
```diff
@@ -125,11 +125,22 @@ def visit_module(mod)
 
         def write_environment_info!(environment, started_at, success)
+          module_deploys = []
+          if environment.puppetfile && environment.puppetfile.modules
```
@Magisus I just realized something while reviewing this:

I believe the `if environment.puppetfile && environment.puppetfile.modules` check is completely unecessary, and could actually cause issues (The intention here is to check to see if the `puppetfile` of `puppetfile.modules` attributes are null, but when testing I ran across the error `environment.puppetfile function is not defined`, actually causing a crash. This was, however when generating mock data to validate data.

Given that each environment has a puppetfile function that returns the puppetfile object, and each puppetfile object has a function that returns its modules, it seems that while this check is unnecessary, it won't actually cause any issues down the line.

I think the proper approach would be to wrap the logic in `write_environment_info!` in a begin/rescue, but I question the value of that give how much overhead is involved. What's your take on it?

Magisus replied:
Looking at those classes, I agree, it's extremely unlikely those things will ever be nil (and this is one of the reasons Ruby's whole mocking paradigm always makes me uncomfortable). I think you're right, given that, it would be clearer to take out. If you put up a PR against 2.6.x, I'll merge it and include it in this mergeup. Good catch!

chrisleicester replied:
awesome! The PR is ready for your review here: https://github.com/puppetlabs/r10k/pull/963

chrisleicester committed:
(CDPE-1813) Modified environment.write_environment_info! to be safer.

Before this change, there existed a slight possibility of the environment.write_environment_into!
method to throw a null pointer exception of a method not found exception under unlikely cicumstances.
This commit wraps the problematic code in a begin/rescue to ensure these problems never occur.

Magisus committed:
(CDPE-1813) Modify environment.write_environment_info! to be sa.. (#963)

(CDPE-1813) Modify environment.write_environment_info! to be safer

Magisus committed:
Merge branch '2.6.x' into 3.0.x

* 2.6.x:
  (CDPE-1813) Modified environment.write_environment_info! to be safer.
  Add test for forge module to .r10k-deploys suite
  (CDPE-1813) Add module deploy info to .r10k-deploy.json file With this change, module information will be added to the output file '.r10k-deploy.json' for external consumption (including the module name, version, and git sha associated with the commit used if pulling the module from a git repository
